cluster/v1alpha1/helpers.go

```diff
@@ -200,28 +215,45 @@ func progressivePerCluster(clusterGroupsMap clusterv1beta1.ClusterGroupsMap, len
 		}
 	}
 
-	clusters := clusterGroupsMap.GetClusters().UnsortedList()
-	clusterToGroupKey := clusterGroupsMap.ClusterToGroupKey()
+	for _, status := range existingClusterStatus {
+		if status.ClusterName == "" {
+			continue
+		}
 
-	// Sort the clusters in alphabetical order to ensure consistency.
-	sort.Strings(clusters)
-	for _, cluster := range clusters {
-		status := statusFunc(cluster)
-		if groupKey, exists := clusterToGroupKey[cluster]; exists {
-			status.GroupKey = groupKey
+		existingClusters[status.ClusterName] = true
+		if status.Status == Succeeded || status.Status == TimeOut {
+			continue
 		}
 
 		newStatus, needToRollout := determineRolloutStatusAndContinue(status, timeout)
 		status.Status = newStatus.Status
 		status.TimeOutTime = newStatus.TimeOutTime
 
+		if status.Status == TimeOut {
+			timeoutClusters = append(timeoutClusters, status)
+			continue
+		}
 		if needToRollout {
-			rolloutClusters[cluster] = status
+			rolloutClusters = append(rolloutClusters, status)
```

Comment:
Need to check the rolloutClusters length I think, should not put all the needToRollout existing clusters into the result.

Reply:
For addon's, there's the possibility that 3 managedclusteraddons are created(existing) and status is "ToApply", if don't check the length, all of the 3 will be returned.

Reply:
I'm assuming the existingClusterStatus hold only clusterStatus; progressing, succeeded, failed and timeout cause the clusterStatus ToApply means the cluster workload does not exist yet. So as addon controller you should not add clusterStatus ToApply to the existingClusterStatus cause the addon (workload) does not exist. What do you think ? 
I will add check to ignore the ToApply clusterStatus at the  existingClusterStatus loop.

Reply:
I think the ToApply means the desired status not applied yet, no matter it's a fresh install (no current workload) or upgrade case (current workload does not match the desired workload) .